predicators/pybullet_helpers/robots/single_arm.py

```diff
@@ -409,6 +409,21 @@ def _ikfast_inverse_kinematics(
             [self.left_finger_joint_idx, self.right_finger_joint_idx])
         final_joint_state.insert(first_finger_idx, self.open_fingers)
         final_joint_state.insert(second_finger_idx, self.open_fingers)
+
+        # Set the robot joints to this final joint state, under the assumptions
+        # that (1) the real robot will move to final_joint_state next and (2)
+        # IK will next be queried with a target state that is nearby the
+        # final_joint_state. This line of code was added to patch a difficult
+        # bug that arose when two versions of the environment were created: one
+        # that exposes the options used by the agent, and one that represents
+        # the "real" environment. In the former environment, IK was getting
+        # called, but the robot's state was not getting updated to match the
+        # state of the "real" environment, leading to subsequent IK failures.
+        # Ideally, none of this would be necessary, but the reality is that
+        # IK is very sensitive to initialization, so it is a stateful method
+        # masquerading as a stateless function.
+        self.set_joints(final_joint_state)
```

Comment:
I understand the explanation, though I would think it should be up to the consumer to set the robot joints to whatever IK gives?

Like here: https://github.com/Learning-and-Intelligent-Systems/predicators/blob/481c2a3f1f599fd49aedae3012332cffbe287898/predicators/pybullet_helpers/robots/single_arm.py#L261-L264

Also, if we modify this for the IKFast I think we need to do the same for the pybullet IK?

Reply:
+1 to modifying pybullet IK too.

about the first part, I agree in principle, but in this case, the consumer is an option policy. it feels quite weird to let the option policy change the robot's state, since the option lives in the mind of the agent but the robot is in the environment...

also, if one day some magical IK solver comes along that is not sensitive to state, we would change this single_arm method, but the consumers shouldn't care...

on the other hand, it's possible that some future consumers would not follow the assumptions that I described in the comment, and so they should be responsible for dealing with that trickiness I guess.

I don't know. damnit IK, why are you so hard.

Reply:
btw if we go the consumer route we'll have to change the kinematic action converter too

Reply:
@williamshen-nz I'm leaning toward keeping this change in the robot itself, mostly to prevent changing multiple consumers. does the name change that I just made help?